common/scala/src/main/scala/org/apache/openwhisk/core/yarn/YARNContainerFactory.scala

```diff
@@ -0,0 +1,476 @@
+  implicit val ec: ExecutionContext = actorSystem.dispatcher
+
+  val runCommand = ""
+  val version = "1.0.0"
+  val description = "OpenWhisk Service"
+
+  var containerCounts = new scala.collection.mutable.HashMap[String, Int]()
+  var containerMap = new scala.collection.mutable.HashMap[String, List[ContainerDefinition]]
+
+  override def init(): Unit = {
+    val images: Set[ImageName] = ExecManifest.runtimesManifest.runtimes.flatMap(a => a.versions.map(b => b.image))
+
+    //Remove service if it already exists
+    if (downloadServiceDefinition() != null)
+      this.cleanup()
+
+    createService(images)
+    waitForServiceStart()
+  }
+
+  def createService(images: Set[ImageName]): Unit = {
+
+    logging.info(this, "Creating Service with images: " + images.map(i => i.publicImageName).mkString(", "))
+    serviceLock.synchronized {
+
+      val componentList = images
+        .map(
+          i =>
+            ComponentDefinition(
+              i.name.replace('.', '-'), //name must be [a-z][a-z0-9-]*
+              0, //start with zero containers
+              runCommand,
+              Option.empty,
+              ArtifactDefinition(i.publicImageName, "DOCKER"),
+              ResourceDefinition(yarnConfig.cpus, yarnConfig.memory),
+              ConfigurationDefinition(Map(("YARN_CONTAINER_RUNTIME_DOCKER_RUN_OVERRIDE_DISABLE", "true")))))
+        .toList
+
+      //Add kerberos def if necessary
+      var kerberosDef: Option[KerberosPrincipalDefinition] = None
+      if (yarnConfig.authType.equals(YARNContainerFactory.KERBEROSAUTH))
+        kerberosDef =
+          Some(KerberosPrincipalDefinition(Some(yarnConfig.kerberosPrincipal), Some(yarnConfig.kerberosKeytab)))
+
+      val service = ServiceDefinition(
+        yarnConfig.serviceName,
+        Some(version),
+        Some(description),
+        "STABLE",
+        Some(yarnConfig.queue),
+        componentList,
+        kerberosDef)
+
+      //Submit service
+      val response =
+        submitRequestWithAuth(HttpMethods.POST, s"${yarnConfig.masterUrl}/app/v1/services", service.toJson.compactPrint)
+
+      //Handle response
+      response match {
+        case httpresponse(StatusCodes.OK, content) =>
+          logging.info(this, s"Service submitted. Response: $content")
+
+        case httpresponse(StatusCodes.Accepted, content) =>
+          logging.info(this, s"Service submitted. Response: $content")
+
+        case httpresponse(_, _) => handleYARNRESTError(response)
+      }
+    }
+  }
+
+  override def createContainer(
+    tid: TransactionId,
+    name: String,
+    actionImage: ExecManifest.ImageName,
+    unuseduserProvidedImage: Boolean,
+    unusedmemory: ByteSize,
+    unusedcpuShares: Int)(implicit config: WhiskConfig, logging: Logging): Future[Container] = {
+
+    logging.info(this, s"Using YARN to create a container with image ${actionImage.name}...")
+
+    Future {
+      serviceLock.synchronized {
```

Comment:
the synchronization here can block the future - does it make sense to synchronize first then run the rest inside the future?

@markusthoemmes might have a better idea.